Tests/NIOTests/ChannelTests.swift

```diff
@@ -1589,7 +1601,66 @@ public class ChannelTests: XCTestCase {
         try serverChannel.close().wait()
     }
 
-    func testEOFOnlyReceivedOnceReadRequested() throws {
+    func testCloseInEOFdChannelReadBehavesCorrectly() throws {
+        let group = MultiThreadedEventLoopGroup(numThreads: 1)
+        defer {
+            XCTAssertNoThrow(try group.syncShutdownGracefully())
+        }
+
+        class VerifyEOFReadOrderingAndCloseInChannelReadHandler: ChannelInboundHandler {
+            typealias InboundIn = ByteBuffer
+
+            private var seenEOF: Bool = false
+            private var numberOfChannelReads: Int = 0
+
+            public func userInboundEventTriggered(ctx: ChannelHandlerContext, event: Any) {
+                if case .some(ChannelEvent.inputClosed) = event as? ChannelEvent {
+                    self.seenEOF = true
+                }
+                ctx.fireUserInboundEventTriggered(event)
+            }
+
+            public func channelRead(ctx: ChannelHandlerContext, data: NIOAny) {
+                if self.seenEOF {
+                    XCTFail("Should not be called before seeing the EOF as autoRead is false and we did not call read(), but received \(self.unwrapInboundIn(data))")
+                }
+                self.numberOfChannelReads += 1
+                let buffer = self.unwrapInboundIn(data)
+                XCTAssertLessThanOrEqual(buffer.readableBytes, 8)
+                XCTAssertEqual(1, self.numberOfChannelReads)
+                ctx.close(mode: .all, promise: nil)
+            }
+        }
+
+        let serverChannel = try ServerBootstrap(group: group)
+            .serverChannelOption(ChannelOptions.socket(SocketOptionLevel(SOL_SOCKET), SO_REUSEADDR), value: 1)
+            .childChannelOption(ChannelOptions.autoRead, value: false)
+            .childChannelInitializer { ch in
+                ch.pipeline.add(handler: VerifyEOFReadOrderingAndCloseInChannelReadHandler())
+            }
+            .childChannelOption(ChannelOptions.maxMessagesPerRead, value: 1)
+            .childChannelOption(ChannelOptions.recvAllocator, value: FixedSizeRecvByteBufferAllocator(capacity: 8))
+            .bind(host: "127.0.0.1", port: 0).wait()
+
+        let clientChannel = try ClientBootstrap(group: group)
+            .connect(to: serverChannel.localAddress!).wait()
+        var buffer = clientChannel.allocator.buffer(capacity: 8)
+        buffer.write(string: "01234567")
+        for _ in 0..<20 {
+            do {
+                try clientChannel.writeAndFlush(buffer).wait()
+            } catch {
+                break
```

Comment:
can you add a comment why its expected that we may see an error ?

Reply:
it shouldn't thanks